File: group_anagrams.cpp
// Memory: 291.3 MB•Time: 4ms•Submitted at: 04/07/2026 10:22
// Beats 100% in speed, 4.26% in memory

#include <algorithm>

using namespace std;

class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        // Create hash map for each anagram group as the key
        // and the index in the result vector as the value

        vector<vector<string>> result;
        map<string, int> indices;
        int i = 1;

        // for each string in strs
        for (string word : strs) {

            // sort the string into temp var
            string sorted = word;
            sort(sorted.begin(), sorted.end());

            // search map for existing entry 
            if (indices.count(sorted) == 0) {

                // if doesnt exist, create entry with temp var and set value to results.size since that is indice of next entry
                indices.insert({sorted, result.size()});

                // add word to result vector at indice
                result.push_back({word});

            } else {
                // otherwise insert at the existing index for the anagram
                int index = indices[sorted];
                result[index].push_back(word);
            }
        
        }

        return result;
    }
};

File: arrays_and_hashing/group_anagrams.py
# Memory: 8.7 MB•Time: 57ms•Submitted at: 04/08/2026 14:18
# Beats 100% in runtime, 20.87% in memory

class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        # Create hash map where
        # key: anagram with letters sorted
        # value: the index in the result vector

        anagram_indices = {}
        anagrams = []
        i = 0;

        # for each string in strs
        for j in range(len(strs)):

            # sort the string into temp var
            og_word = strs[j]
            sorted_word = "".join(sorted(str(og_word)))

            # search map for existing entry 
            if sorted_word not in anagram_indices:

                # if doesnt exist, create entry with temp var and set value to i
                anagram_indices[sorted_word] = i

                # create new entry in dict for the word
                anagrams.append([og_word])

                # increment i
                i += 1
            else:
                # otherwise append to appropriate list in dict
                anagrams[anagram_indices[sorted_word]].append(og_word)

        return anagrams
